### PR TITLE
Diego QR Code App

### PR DESCRIPTION
**Pull request by diegogarciar**

### CODE REVIEW

File: lib/infrastructure/dtos/seed_dto.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  const SeedDto._();
+
+  const factory SeedDto({
+    required String seed,
+    @JsonKey(name: 'expires_at') required String expiresAt,
+  }) = _SeedDto;
+
+  Seed toDomain() {
+    return Seed(seed: seed, expiresAt: DateTime.parse(expiresAt));
+  }
```
Comment (russell-sf) [inline lib/infrastructure/dtos/seed_dto.dart:18]: **suggestion**: The Dto class should have one responsibility. That is serializing data to and from json. The conversion of data to domain layers can be handled by its own Mapper class.

File: lib/presentation/custom_widgets/fab_button/expandable_fab.dart [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+          progress: CurvedAnimation(
+            curve: Curves.fastOutSlowIn,
+            reverseCurve: Curves.easeOutQuad,
+            parent: animation,
+          ),
+          child: this.children[i],
+        ),
+      );
+    });
+    return children;
```
Comment (russell-sf) [inline lib/presentation/custom_widgets/fab_button/expandable_fab.dart:61]: **suggestion**: There is a little cleaner way to do this...

```suggestion
    return children.asMap().keys.map((index) {
        return _ExpandingActionButton(
          index: index,
          progress: CurvedAnimation(
            curve: Curves.fastOutSlowIn,
            reverseCurve: Curves.easeOutQuad,
            parent: animation,
          ),
          child: children[index],
        ),
      );
    }).toList();
```

File: lib/infrastructure/seed_generator_repository.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+@LazySingleton(as: ISeedGeneratorRepository)
+class SeedGeneratorRepository implements ISeedGeneratorRepository {
+  final SharedPreferences _sharedPreferences;
+  final Dio _dio;
+
+  static final endpoint = Uri.parse('http://mtask.mx/seed');
+  static const cacheKey = "seed";
+
+  SeedGeneratorRepository(
```
Comment (russell-sf) [inline lib/infrastructure/seed_generator_repository.dart:21]: **suggestion**: This could be const.

**Review by russell-sf — COMMENTED** [review]
Hi Diego! Thank you for your code submission. I've gone over it and just have a few minor comments.

I really appreciate the clean and easy to follow code and the attention to detail. I really cannot find major faults with what you have done.  Great job!

If you have any questions regarding my feedback then leave comments.

File: lib/presentation/custom_widgets/fab_button/expandable_fab.dart (hunk 2 above)
Comment (diegogarciar) [reply lib/presentation/custom_widgets/fab_button/expandable_fab.dart:61]: Cleaner indeed!

File: lib/infrastructure/seed_generator_repository.dart (hunk 3 above)
Comment (diegogarciar) [reply lib/infrastructure/seed_generator_repository.dart:21]: You actually got me here, how can this be done? As my understanding Uri.parse does not have const constructors.
```
Const variables must be initialized with a constant value.
Try changing the initializer to be a constant expression.
```

and as for cacheKey I'm certain I cannot have them inside classes or else I get the following error: 
```
Only static fields can be declared as const.
Try declaring the field as final, or adding the keyword 'static'.
```
Do you mean declaring cacheKey outside the class? Like a file named 'constants.dart' with all my constants and importing it

**Comment by diegogarciar** [comment]
Thank you for the review and the feedback @russell-sf!

File: lib/infrastructure/seed_generator_repository.dart (hunk 3 above)
Comment (russell-sf) [reply lib/infrastructure/seed_generator_repository.dart:21]: Sorry I was referring to the constructor being const.

**Review by jeffs-sf — COMMENTED** [review]
Hello @diegogarciar, thank you for taking the time to complete this submission. Overall this looks solid, and I agree with @russell-sf that there are no major faults. 

I really appreciate the organization, testing, and comments that are present throughout.

One thing I noticed that seemed a little weird was this behavior below, any ideas what would be causing that? I seemed to be able to reproduce it by scanning a QR Code, moving off the QR Code, and then scanning the same one again. Notice the glitchy/flashing behavior:

https://user-images.githubusercontent.com/53348890/124649897-1168a300-de67-11eb-9a83-86eff150afdd.MP4

**Comment by diegogarciar** [comment]
Hello @jeffs-sf,
Thank you for the positive feedback, I really enjoy working with Flutter and I'm glad to know I'm doing decent.

For the glitch, my guess is that as I don't stop scanning, it is constantly "validating" the "new" scanned text. As I would like to keep this constant validation feature, I would just validate the seed if the new QR scanned is the first one scanned or if its different from the last.

**Comment by jeffs-sf** [comment]
Thank you again for your submission and it was a pleasure speaking with you.

**Pull request closed without merging**